lib/Optimizer/TrainingPreparation.cpp

```diff
@@ -0,0 +1,118 @@
+      tensor->init(Tensor::InitKind::Broadcast, 0.1, F->getPRNG());
+    }
+  } break;
+  case Kinded::Kind::BatchNormalizationNodeKind: {
+    BatchNormalizationNode *BN = llvm::dyn_cast<BatchNormalizationNode>(node);
+    if (BN->getScale() == input) {
+      tensor->init(glow::Tensor::InitKind::Zero, 0, F->getPRNG());
+    } else if (BN->getBias() == input) {
+      tensor->init(glow::Tensor::InitKind::Broadcast, 0.1, F->getPRNG());
+    } else if (BN->getMean() == input) {
+      tensor->zero();
+    } else if (BN->getVar() == input) {
+      tensor->init(glow::Tensor::InitKind::Broadcast, 1.0, F->getPRNG());
+    }
+  } break;
+  case Kinded::Kind::FullyConnectedNodeKind: {
+    FullyConnectedNode *FCN = llvm::dyn_cast<FullyConnectedNode>(node);
+    if (FCN->getWeights() == input) {
+      auto in = FCN->getInput();
+      tensor->init(glow::Tensor::InitKind::Zero, in.dims()[1], F->getPRNG());
+    } else if (FCN->getBias() == input) {
+      tensor->init(glow::Tensor::InitKind::Broadcast, 0.1, F->getPRNG());
+    }
+  } break;
+  default:
+    break;
+  }
+}
+} // namespace
+
+TensorInitializer getDefaultTensorInitializer() {
+  return defaultTensorInitializer;
+}
+
+void prepareFunctionForTraining(
+    Function *F, PlaceholderBindings &bindings,
+    const std::unordered_set<std::string> &legalOutputNames,
+    TensorInitializer &&initializer) {
+
+  auto &nodes = F->getNodes();
+  // Lookup all nodes, skip Storage types, enumerate inputs,
+  // replace Constant type with trainable Placeholders except special cases,
+  // like BatchNormalization inputs (mean and variance). In special cases
+  // replace Constant type with non-trainable Placeholders.
+  for (auto &node : nodes) {
+    // Skip storages.
+    if (Storage::classof(&node)) {
+      continue;
+    }
+
+    // Skip output node(s).
+    if (legalOutputNames.count(node.getName())) {
+      continue;
+    }
+
+    for (unsigned idx = 0; idx < node.getNumInputs(); ++idx) {
+      auto INV = node.getNthInput(idx);
+      auto *IN = INV.getNode();
+      if (Constant *C = llvm::dyn_cast<Constant>(IN)) {
+        bool isTrainable = true;
+        switch (node.getKind()) {
+          // Special case. Mean and Variance inputs are not trainable.
+        case Kinded::Kind::BatchNormalizationNodeKind: {
+          if (BatchNormalizationNode *BN =
+                  llvm::dyn_cast<BatchNormalizationNode>(&node)) {
+            if (INV == BN->getMean() || INV == BN->getVar()) {
+              isTrainable = false;
+            }
+          }
+        } break;
+        default:
+          break;
+        }
+        auto *PH = F->getParent()->createPlaceholder(C->getType(), C->getName(),
+                                                     isTrainable);
+        C->getOutput().replaceAllUsesOfWith(PH, F);
+        bindings.insert(PH, std::move(C->getPayloadMutable()));
+        F->getParent()->eraseConstant(C);
+        auto *tensor = bindings.get(PH);
```

Comment:
I *think* you can avoid this extra lookup if you initialize the tensor before inserting it into `bindings`:
```
auto &tensor = C->getPayloadMutable();
initializer(F, &node, INV, tensor);
bindings.insert(PH, std::move(tensor));
F->getParent()->eraseConstant(C);
```